### PR TITLE
build.rb: Add dependencies to RPATH for Linuxbrew

### PR DESCRIPTION
**Pull request by sjackman**
Add opt_lib of each dependency to RPATH.

### CODE REVIEW

**Comment by sjackman** [comment]
This patch adds the `opt_lib` directory of each dependency to the `RPATH` via `LDFLAGS` and `LD_RUN_PATH`. The first entry of the `RPATH` remains `$HOMEBREW_PREFIX/lib`, so these directories are used as a fall-back if a formula is not linked. It's necessary for keg-only libraries, which we didn't have any of in Linuxbrew until now with the versioned libraries, such as `gmp@4`. The `RPATH` is necessary for these keg-only libraries (see e.g. https://github.com/Linuxbrew/homebrew-core/pull/1574). Two design questions must be answered:

1. Should the `RPATH` include all dependencies or only keg-only dependencies? I've opted for all dependencies. This allows an executable to continue to work even while one if its dependents has been unlinked, for example `curl` will continue to work through `brew upgrade openssl`.
2. Should `$HOMEBREW_PREFIX/lib` be first or last? I've opted for first. This retains the current behaviour and gives the user control over dynamic linking by modifying the symlinks in `$HOMEBREW_PREFIX/lib`.

File: Library/Homebrew/build.rb [hunk 1]
```diff
@@ -100,6 +100,15 @@ def install
         ENV.prepend "LDFLAGS", "-L#{dep.opt_lib}" if dep.opt_lib.directory?
         ENV.prepend "CPPFLAGS", "-I#{dep.opt_include}" if dep.opt_include.directory?
       end
+
```
Comment (bw-yuhogg) [inline Library/Homebrew/build.rb:105]: I think a comment explaining the rationale for this is warranted here.

File: Library/Homebrew/build.rb [hunk 2]
```diff
@@ -100,6 +100,15 @@ def install
...
+      if OS.linux?
+        formula_deps.each do |dep|
+          ENV.append_path "LD_RUN_PATH", dep.opt_lib
+        end
+        unless ENV["LD_RUN_PATH"].empty?
+          ENV.prepend "LDFLAGS", "-Wl,-rpath,#{ENV["LD_RUN_PATH"]}"
```
Comment (bw-yuhogg) [inline Library/Homebrew/build.rb]: Minor thing: /home/linuxbrew/.linuxbrew/lib may show up twice in the rpath after doing this.

As a test:

```
% brew install -s mksh monit
% patchelf --print-rpath /home/linuxbrew/.linuxbrew/bin/mksh
/home/linuxbrew/.linuxbrew/lib
% patchelf --print-rpath /home/linuxbrew/.linuxbrew/bin/monit
/home/linuxbrew/.linuxbrew/lib:/home/linuxbrew/.linuxbrew/lib:/home/linuxbrew/.linuxbrew/opt/zlib/lib:/home/linuxbrew/.linuxbrew/opt/openssl/lib
```

Comment (sjackman) [reply Library/Homebrew/build.rb]: That's not ideal, but also won't break anything. That might happen if the tool's build system also uses a `-Wl,-rpath` option.

**Review by bw-yuhogg — COMMENTED** [review]
I'm not entirely sure why we need to use both -Wl,-rpath,$LD_RUN_PATH and set LD_RUN_PATH explicitly.

Assuming we're only supporting Clang and GCC, plus their corresponding linkers, shouldn't one or the other be enough? https://www.eyrie.org/~eagle/notes/rpath.html

**Comment by sjackman** [comment]
> I'm not entirely sure why we need to use both -Wl,-rpath,$LD_RUN_PATH and set LD_RUN_PATH explicitly.

We set `LDFLAGS`, but some build systems ignored `LDFLAGS`, and  `LD_RUN_PATH` is our backup if `LDFLAGS` is ignored. We could set only `LD_RUN_PATH`, but I like the verbosity of `LDFLAGS` appearing in the logs and command lines. `--env=super` can resolve this ugliness.

**Pull request merged**